### ISSUE TITLE
How to monetize NodeOS?

### ISSUE THREAD

**Issue by piranna** [NodeOS/NodeOS#279]
I have been thinking about this topic since a long time ago but didn't got yet a clear solution. It seems NodeOS has got some interest to be used on production or it's already being used for it, but for being this possible in the long term it would need to be dedicated some more time and not being only a "hobby" done in our spare time both for maintenance and development of new features (NodeOS has became a really big project with a lot of dependencies that need to be maintained, although I'm trying to automate these tasks the most I can), and since we are several here that would be interested on getting some revenues, I think it's time to ask the big question: how to monetize NodeOS, so I've open this issue to discuss for alternatives and try to find solutions for this.

The main problem I think we have are:

- How to measure contributions
- There are collaborators from around the world in different timezones
- There are collaborators that don't participate anymore

This makes it difficult to define a model fair for everybody beyond to use BountySource where anybody gets payed for the issues that he solve, but that doesn't give promotion of NodeOS at big scale. It would be possible to create a company/association (I think Germany laws are easy for this, maybe @michielbdejong can give us some advice here, could you? :-) ) that give support, consultancy and expertise about NodeOS, but here the problem is how don't give the feel that NodeOS is owned by that company.

A NodeOS Foundation would be the coolest option by putting the project in the first place and maybe would make it easier to get some sponsors and assists to events representing to NodeOS and not only just as individuals (as I have done on the keynotes and championships, but try to give on them the attributions to @groundwater for starting the project and the other collaborators as much as I can), but here we come again the the first problem: how to distribute the money in a fair way.

Another option, maybe the better, is to forget by the moment about the beneficts and create a non-proffit organization just to have some identity and to promote NodeOS itself, try to get some sponsors for travelling to conferences and little more (at this moment we don't have any cost on servers or similar, only the website domain), and if things go well, maybe promote the organization to a foundation and dedicate those sponsor funds to pay a developer for one or two months working full or part-time or by objectives. I admit from start that I would candidate for that if we get to that situation (just to be clear and honest), but although I'm the most active one here and has a somewhat mind-map of how I want NodeOS to be, I'm not propossing this to "get all the money" but instead I'm very concerned and worried about how to make others to benefict for their work on NodeOS.

Also, there's the big problem of promote NodeOS and nock some doors to get sponsors of the project and financiation... I think Node.js Foundation and NPM org would be the obvious candidate to ask for, but we would need to show we are not just some kids playing here, so adding native git support for npm with a pure javascript library instead of relaying on an external git cli command would be a really good presentation letter. Later we would consider other big players like Mozilla Foundation if we get a good integration with B2G OS (I know it's community driven, but sure they are not unaware what's happening on their community ;-) ) and if we got luck we could also ask to the Linux Foundation as an example of how powerful and easy to use Linux has became in the last years.

Opinions? Suggestions? There are some list here on GitHub about finatial mechanism for Open Source projects, it would be a good idea to reference them here too.

/cc @luii, @5paceManSpiff, @SpaceboyRoss01, @kskarthik, @Coretool, @lapineige, @lissyx, @Ryuno-Ki, @khanshakeeb

**Comment by dryamJS** [comment]
Don't give up, you project is amazing! It is a great idea. I think that you can start from donation button on the blog pages. P.S. I'am sorry for my poor english.

**Comment by luii** [comment]
> [...] so adding native git support for npm with a pure javascript library instead of relaying on an
> external git cli command would be a really good presentation letter.

Like i mentioned on #67 im writing `nogit` (nodeos-git) using `nodegit` which is a wrapper for `libgit2` and currently im implementing the features for the 1.0.0 Milestone.

Unlikely i dont have much time but currently im ill and can do some progress on `nogit`.

**Comment by piranna** [comment]
> I think that you can start from donation button on the blog pages

Yeah, but where does this money (if any) goes? It would not be fair to go just to my pocket, and NodeOS don't have a legal figure to send it there. Each one can have a "pay me and I'll work on NodeOS" button, but what developer should I pay then? We are several here, and it would make people loose the focus on the project...


> Like i mentioned on #67 im writing nogit (nodeos-git) using nodegit which is a wrapper for libgit2 and currently im implementing the features for the 1.0.0 Milestone.

I know, this was just an example of how we can get their attention :-) Or maybe we could ask them for money in advance to do it, too... Why don't you try to ask them?

**Comment by Coretool** [comment]
@piranna We could monetize support for commercial projects (and maybe solve it like Qt ?). 

A clever way would be solving distribution via shares so who owns most shares gets most (in this case probably you because you're kind of the lead dev). We could distribute shares on the devs and set of a part of the shares to be bought (so we get a kind of starting capital). Later we would gain via the licence and profit would be distributed via shares. That's the way silicon valley solved it ...

**Comment by aulvi** [comment]
@piranna You say that you wish to monetize NodeOS but I seem to have miss the important part: why do you need money?

Personally, I think the best approach at this stage would be to promote the project like crazy. Recruit contributors, build a healthy project. The contributor's graph shows that this is pretty much a one-person project.

That leads me to my next point, paying developers to support an OSS project. One risk is that when the money dries up, the project dies. Additionally, it can be difficult to maintain project autonomy since there is now a financial incentive to keep your sponsors happy. Allowing the project to bend to the will of sponsors is risky because it may harm your ability to secure funding from other sources.

If, for example, you get a check from Mozilla, would you suddenly make B2G a higher priority? If so, how would other potential sponsors feel about that?

Many foundations serve to help projects stay "pure" by keeping the funding separate from the development effort. I believe that benefit alone is worth consideration.

Anyways, this topic is pretty huge and I'm certainly curious about where things go from here.

**Comment by vitkarpov** [comment]
> Recruit contributors, build a healthy project.

A really good point! For instance, I'd like to participate, but I really don't understand where to start. Are there any talks, docs, blog posts about the main idea, architecture overview, use cases on so on and so forth?

**Comment by lite20** [comment]
@vitkarpov the readme on the main page links to our current two "todo lists". We probably should link to where progress on these is being done but for now if anything peaks your interest just let me know. A new wiki is under works but there is plenty of information available to help understand how things work in the issue threads and in the current github wiki.

Should you have any questions, just tag me and I'll try to help. Welcome to NodeOS!

**Comment by piranna** [comment]
> @piranna You say that you wish to monetize NodeOS but I seem to have miss the important part: why do you need money?

Are you asking on a personal point of view, or about NodeOS? In the last case, the main reason is to be able to "work" on NodeOS and be able to give it the attention that's requiring now that it's getting so big, where other jobs and personal life don't allow me or others to give it time, delegating NodeOS to being it just a hobbie and maybe losing the opportunity to make it something bigger to what it's now...


> Personally, I think the best approach at this stage would be to promote the project like crazy. Recruit contributors, build a healthy project.

This is a good idea too. How could this be done in an effective way? Do you have some ideas? So far I have only managed to do some keynotes... Some companies got interest on NodeOS but only about to use it as a final product.


> The contributor's graph shows that this is pretty much a one-person project.

And it's not something I'm fully proud about... One of the things that need to be done is improve documentation to make it easier to newcomers to add new contributions, specially graph diagrams, how-tos, how components are related ones to others...


> That leads me to my next point, paying developers to support an OSS project. One risk is that when the money dries up, the project dies. Additionally, it can be difficult to maintain project autonomy since there is now a financial incentive to keep your sponsors happy. Allowing the project to bend to the will of sponsors is risky because it may harm your ability to secure funding from other sources.
> 
> If, for example, you get a check from Mozilla, would you suddenly make B2G a higher priority? If so, how would other potential sponsors feel about that?

I don't find it so bad as far as the sponsors only give suggestions or require addons and don't interfere with the long term project design moving it at their will, but this is just my personal opinion...


> Many foundations serve to help projects stay "pure" by keeping the funding separate from the development effort. I believe that benefit alone is worth consideration.

To have unrelated to get the funds and to code the project? Yes, that would be ideal, specially because I'm not a public relationship kind of person. Seems Node.js itself works this way, having the Node.js Foundation to replace the funds management of StrongLoop, and without its influence on the code... This is somewhat related what I told before about creating an organization or foundation.


> Anyways, this topic is pretty huge and I'm certainly curious about where things go from here.

Me too, I opened it also to start the discussion, thank you for your comments :-)


> Recruit contributors, build a healthy project.
> 
> A really good point! For instance, I'd like to participate, but I really don't understand where to start. Are there any talks, docs, blog posts about the main idea, architecture overview, use cases on so on and so forth?

In this way, probably the most complete documentation is [my degree thesis](https://github.com/piranna/pfc). It's released as CreativeCommons and in GitBook format but it's on spanish, I'm open for contributions if someone want to update or translate it, but please first tell me it so I can be able to add support for several languajes (or add support for it first yourself :-) ).


> @vitkarpov the readme on the main page links to our current two "todo lists". We probably should link to where progress on these is being done but for now if anything peaks your interest just let me know. A new wiki is under works but there is plenty of information available to help understand how things work in the issue threads and in the current github wiki.

A page on the wiki showing the dependencies tree and links to all the NodeOS related projects would be good too... :-)

**Comment by Ryuno-Ki** [comment]
@piranna [wrote](https://github.com/NodeOS/NodeOS/issues/279#issue-169331582):
> I think Node.js Foundation and NPM org would be the obvious candidate to ask for

What about the jQuery Foundation?

@Coretool [wrote](https://github.com/NodeOS/NodeOS/issues/279#issuecomment-237538874):
> @piranna We could monetize support for commercial projects (licences, one time fee) and maybe solve it like Awesomium ?

Aurelia solved it by offering customization for money. Themes and that like.

@piranna [wrote](https://github.com/NodeOS/NodeOS/issues/279#issuecomment-237710532):
> In this way, probably the most complete documentation is my degree thesis. It's released as CreativeCommons and in GitBook format but it's on spanish, I'm open for contributions if someone want to update or translate it, but please first tell me it so I can be able to add support for several languajes (or add support for it first yourself :-) ).

English would be a good starter (lingua franca to most people). Spanish is widely used also.

@piranna [wrote](https://github.com/NodeOS/NodeOS/issues/279#issuecomment-237710532):
> A page on the wiki showing the dependencies tree and links to all the NodeOS related projects would be good too... :-)

Reminds me to get [grunt-graphviz](euskadi31/grunt-graphviz) running again … if time permits …

**Comment by aulvi** [comment]
@piranna Great response, thanks for getting this Issue going, I'm enjoying the discussion.

Okay, so really what you're after at this stage is development resources. Making it easy for someone to contribute is the first step, and often that means writing documentation and crafting a solid development experience. Following that, the Issues should be accessible as well, using labels to identify estimated effort and difficulty can help new contributors ease their way into the project.

Once it's easy to add collaborators then you can recruit for help without getting overwhelmed, and without the new collaborating getting frustrated and wandering off.

Question: outside of github issues, where can we chat? I'd love to just catch up one-on-one on IRC or Slack or what-have-you.. how can this be done?

**Comment by formula1** [comment]
In terms of documentation I think most pf it structurel vs api. Something like JSDoc will be useless considering most of the important are things like filestructure, user organization, startup and shell operations. I know groundwater explained asguard (though hes no longer with us) but ultimately that was high level with minimal to no hooks or events in between.

Whats the events that can / should be plugged into? How do you plug into this events via "installation"? How can I help @piranna? :)

**Comment by Ryuno-Ki** [comment]
@aulvi [wrote](https://github.com/NodeOS/NodeOS/issues/279#issuecomment-237983700):
> Question: outside of github issues, where can we chat? I'd love to just catch up one-on-one on IRC or Slack or what-have-you.. how can this be done?

What about Gitter? It should be fairly easy to create a gitter.im channel for this repository. You can log in via GitHub their.

**Comment by Coretool** [comment]
@Ryuno-Ki We have gitter.

**Comment by Ryuno-Ki** [comment]
@Coretool I see.

@aulvi Have a look at https://gitter.im/NodeOS/NodeOS

**Comment by piranna** [comment]
@Ryuno-Ki:
> What about the jQuery Foundation?

Any help would be welcome :-)


> Aurelia solved it by offering customization for money. Themes and that like.

Not bad idea, that's mostly what flavours are intended for... Maybe should I put a list of this ideas in the first post for reference and don't need to dig all the messages?


> English would be a good starter (lingua franca to most people). Spanish is widely used also.

I did it in spanish because they told me doing it in english don't give any extra value, and also since it's not my mother language, it would be more difficult to be reviewed by others and if I have had some grammar failures it would be against my final score... :-/ Anyway, I graduated with distinction, so what the f*ck :-P

I have upgraded it to add support for multilanguage so we could start to translate it (english first, of course), but the book generation is failing due to the [mermaid plugin](https://github.com/JozoVilcek/gitbook-plugin-mermaid) and the usage of PhantomJS, so I think I'll need to change it for [another one that don't make use of PhantomJS and also it's more updated](https://github.com/yanni4night/gitbook-plugin-mermaid-full), but would need to add him the support of load mermaid files from the filesystem too... :-/ Any volunteer? :-D


@aulvi:
> @piranna Great response, thanks for getting this Issue going, I'm enjoying the discussion.

You are welcome, that was my intention :-)


> Okay, so really what you're after at this stage is development resources.

Mostly yes, we need more horse power here, being that more contributions and just only need to accept pull-requests and responding emails, or be able to have one or several developers working part or full-time on the project, specially for the "big things" like refactoring the build mechanism to use the QEmu nomenclature. Add to this the fact that I have been having a really strange and bad year both personal professional and economically (is that the 30 years old crisis?) needing to pick some frustrating jobs to pay the bills and removing hours of sleep to dedicate them to NodeOS up to the point of thinking about to f*ck off everything just after finishing my degree and just code by hobbie on my spare time, and you have the perfect storm to spontaneouly think about how you would make a life from the project where you really want to invest time, and that in fact needs someone to invest time in it too. Point is, having so much people involved here, it's not easy to don't make people think you are an *sshole... :-/


> Making it easy for someone to contribute is the first step, and often that means writing documentation and crafting a solid development experience. Following that, the Issues should be accessible as well, using labels to identify estimated effort and difficulty can help new contributors ease their way into the project.

I agree on that, and hope we can move towards this. The [docs repo](https://github.com/NodeOS/docs) is the correct place to put this, question is, what should a newcomer expect from NodeOS, and how we can guide him to add contributions? I think in that way to document the dependencies tree go in the correct path, and a CONTRIBUTE.md file would help too. The issues labels is a good thing too, we are underusing them but they are not easy to create on the fly...


> Question: outside of github issues, where can we chat? I'd love to just catch up one-on-one on IRC or Slack or what-have-you.. how can this be done?

I'm not too much into chats due to interactivity and too much noise (I use WhatsApp, and thank you), I like more email and/or GitHub issues and answer when I have some spare time, but since Slack is getting some traction, maybe we could create one if there's not one already... I think there was a Gitter and some people were using it, but as I've said, I'm not sure because I'm not into chats :-P Anyone can confirm that?


@formula1 
> In terms of documentation I think most pf it structurel vs api. Something like JSDoc will be useless considering most of the important are things like filestructure, user organization, startup and shell operations. I know groundwater explained asguard (though hes no longer with us) but ultimately that was high level with minimal to no hooks or events in between.
> 
> Whats the events that can / should be plugged into? How do you plug into this events via "installation"? How can I help @piranna? :)

I think a good start would be to add some text explaining what the modules do and how in plain english on the README.md files of the different modules repos. That could be done by reviewing the source code of them, that can be used to add JsDoc to them and also upgrade to ES6 or fix some bugs if someone is into the mood :-) Adding tests can help too, not only by finding failures on corner cases but also to think about how the APIs would be made easier and cleaner... everybody wins :-)


> What about Gitter? It should be fairly easy to create a gitter.im channel for this repository. You can log in via GitHub their.

If it's not already done and you want to create and maintain it, go for it :-)

**Comment by Ryuno-Ki** [comment]
@piranna [wrote](https://github.com/NodeOS/NodeOS/issues/279#issuecomment-238076522):
> I have upgraded it to add support for multilanguage so we could start to translate it (english first, of course), but the book generation is failing due to the mermaid plugin and the usage of PhantomJS, so I think I'll need to change it for another one that don't make use of PhantomJS and also it's more updated, but would need to add him the support of load mermaid files from the filesystem too... :-/ Any volunteer? :-D

Hey, @kurai021 - you've worked with Mermaid back then. Can you have a look into this?

@piranna [wrote](https://github.com/NodeOS/NodeOS/issues/279#issuecomment-238076522):
> The issues labels is a good thing too, we are underusing them but they are not easy to create on the fly...

I wonder how [Greenkeeper does it](https://greenkeeper.io/verify.html) …

Concerning Gitter look at https://github.com/NodeOS/NodeOS/issues/279#issuecomment-238076120

**Comment by piranna** [comment]
I have just added some level tags to the issues and filled the issues with them. They are

- **easy**: simple tasks that a newcomer would do it without too much problems or that can be done in short time
- **medium**: easy tasks, but needs to know a bit how NodeOS works or it's structured
- **hard**: really need to know what you are doing, mostly add new features. Probably several hours working on several projects simultaneously
- **gurú: needs to have a global view of how NodeOS works and where it moves fordward, and/or usually they are needed knowledges on systems programming. Usually they would need several hours or days to fully achieve them

I've also added a tag for the task that would be long to achieve and probably would be better to collaborate with others to achieve it.

**Comment by piranna** [comment]
They have just told me about [Precipita](http://www.precipita.es/) initiative, a crowdfunding platform for scientific projects that we would be able to opt-in. There are two requeriments that needs to be fullfilled by someone of the group, work in a public organism, and have some scientific publications in the last two years. I'm working in the university, but don't have any article yet, is there someone that could fullfill the requeriments?

On the other hand, I have been thinking some time ago that I could be able to write some papers about NodeOS technology and infraestructure, and in fact this would not only help to promote the project and the ones that help to write the paper, but also would help as documentation of the project :-) Anyone is interested to join me on this? :-D

**Comment by Coretool** [comment]
I maybe know somebody who would could opt as the guy with the publication ... He's very interested in your / our work on NodeOS ;) 

Maybe I could also help writing ...

**Comment by piranna** [comment]
Any help is welcome :-)

**Comment by piranna** [comment]
Any help is welcome :-)

**Comment by piranna** [comment]
Mozilla awards are open until end of month :-D

https://blog.mozilla.org/blog/2016/08/04/mozilla-awards-585000-to-nine-open-source-projects-in-q2-2016/

**Comment by piranna** [comment]
Shoul we participate? :-)

**Comment by kurai021** [comment]
@Ryuno-Ki I have worked with Mermaid.js and Markdown-pdf to generate documents, but Markdown-pdf depends of Phantom.js. What's the problem here?

**Comment by Coretool** [comment]
@piranna If we can't lose something, why not ?

**Comment by luii** [comment]
We cant lose anything, why not.

**Comment by piranna** [comment]
> @Ryuno-Ki I have worked with Mermaid.js and Markdown-pdf to generate documents, but Markdown-pdf depends of PhantomJS. What's the problem here?

The problem seems to be related to `gitbook-plugin-mermaid` using an old version of PhantomJS. I think the best option would be to use one of the alternative modules that don't require it, but they would need some improvements to add support to load Mermaid diagrams from the filesystem instead of being hardcoded...

**Comment by piranna** [comment]
https://wiki.mozilla.org/MOSS/Mission_Partners
> The project’s home can be anywhere in the world where we can make payment without undue burden. We have a reasonably strong preference for funding legal organizations rather than individuals representing projects, although exceptional circumstances could lead to some flexibility.

I supposed it... We would need to think what we do with that... :-/

Here you have the inscription form if you want to provide comments, but questions like "where to send the money" and "how much do you need" are complicated to answer... :-/ Probably it's better to try to organize ourselves and work on generate a good documentation and create a healthy community first. On the other hand, is there someone with experience creating associations or organizations? I only had it as vocal, no more...

**Comment by aulvi** [comment]
Community is a necessity even if we were well-funded, so I believe that's a good place to start.

We need contributors. Let's reach out on Twitter, friendly blogs, tech sites, etc and make it known that we're looking for folks to help at all levels. I'd be more than happy to help with this part.

**Comment by michielbdejong** [comment]
>  It would be possible to create a company/association (I think Germany laws are easy for this, maybe @michielbdejong can give us some advice here, could you? :-) ) that give support, consultancy and expertise about NodeOS, but here the problem is how don't give the feel that NodeOS is owned by that company.

IMHO, governments and the legal entities they offer are old-fashioned. Sure, everybody who receives money from a donations pot in exchange for (or as an enabler for) software development or other work should write that money onto their tax statement in the country where that person happens to pay taxes. But you can write it down as 'income from other activities', and your government will tax you appropriately.

Other than that, you don't need the involvement of any nation state government, you can just keep track of donations and project finance using a text file, spreadsheet or database. It's not necessary to have a bank account in the project's name to do things like crowd-funding, flattr, bountysource, etc. You can just link a contributor's personal paypal or bank account. I would say avoid paperwork, it will just distract and cost money. :)

**Comment by junland** [comment]
>Community is a necessity even if we were well-funded, so I believe that's a good place to start.

>We need contributors. Let's reach out on Twitter, friendly blogs, tech sites, etc and make it known that we're looking for folks to help at all levels. I'd be more than happy to help with this part.

Agreed, focus on the community (Your already doing this and it's doing very well.) when you put community first the money / funding will follow, one way or another. Just keep your heads up and keep making this project awesome.

Just wanted to chime in :)

**Comment by piranna** [comment]
@aulvi
> Community is a necessity even if we were well-funded, so I believe that's a good place to start.

It's clearly that at this stage the first move is to create a healthy community and try get NodeOS to live by itself without depending of living-assistance funds, so let's focus on it. I have already change the issue topic to *How to get more contributors?* to show this, sorry if someone would be offended for the previous title or the initial intention of the issue to raise money for the project and/or its developers/contributors.


@michielbdejong
> Other than that, you don't need the involvement of any nation state government, you can just keep track of donations and project finance using a text file, spreadsheet or database. It's not necessary to have a bank account in the project's name to do things like crowd-funding, flattr, bountysource, etc. You can just link a contributor's personal paypal or bank account. I would say avoid paperwork, it will just distract and cost money. :)

Hi dude! Please to read you again! :-) So, are you saying that's enough by having a public doc where people can see what money goes in and out so everybody can see that things are transparent and fair?


> do some freelance work to save up some money, then take a few months "sabbatical" to work on your open source project. When people see on the project website that you're a main contributor, you could add a link saying 'Available for hire' and get paid gigs like that

Willing to live by coding free software, both that advices are really interesting from a personal point of view :-) Thanks dude! :-D


@junland
> Agreed, focus on the community (Your already doing this and it's doing very well.)

Thank you! ^^ I think I can do it better, but thank you anyway :-)

**Comment by aulvi** [comment]
@piranna Awesome stuff, thanks again for getting this Issue moving forward. Please let me know how I can help spread the word.

**Comment by piranna** [comment]
> @piranna Awesome stuff, thanks again for getting this Issue moving forward. Please let me know how I can help spread the word.

* Main thing, start moving it on social networks like Twitter and Facebook.
* Local Meetups and conferences you would already assists or know (specially the Node.js related ones) are a good place to meet people that would be interested on the project, and having people talking about NodeOS on [Node.js Interactive Europe](http://events.linuxfoundation.org/events/node-interactive-europe) this September would be really cool... :-D That's why I talked about having some minimal funds for travelling expenses to conferences and so :-)
* Universities and computer clubs, and systems department teachers would be interested too.
* Maybe at [OsDev](http://www.osdev.org/) and other similar forums would be interested to take a look...

But maybe the most important thing would be to take a look at the project as a newcomer and each one ask himself: "I don't know nothing about the project but want to contribute, what info I would like to find to start poking around?"

**Comment by piranna** [comment]
I have created yesterday the [NodeOS project dependencies tree](https://github.com/NodeOS/NodeOS/wiki/NodeOS-project-dependencies-tree) page on the wiki, so we now can have a high-fly view of how NodeOS is organized... and it's huge :-P

I have included all the dependencies expanding them up to the point where it's found an external project (should we remove them and left only "useful" links?), and put links and badges (where available) to projects of NodeOS organization or done by NodeOS contributors for them, being in *cursive* the ones that are unmerged forks of third-party projects. It lacks the projects that are not directly involved with NodeOS itself, like the web page or [coverdeeps](https://github.com/piranna/coverdeeps).

So from this first iteration, things that I can see can be done:

- [ ] we have a lot of projects without (complete) tests and also without a `README.md`, mostly due to my fault to try to get them working fast to have NodeOS ready for the championships, so adding them and configuring a CI server to check tests coverture (and add the badges :-D ) would be a good task for a newcomer that want to get into NodeOS source code :-) Let's collect as much (useful) badges as we can and make that wiki page seem like a Christmas tree! :-D
- [ ] get to be [send](https://github.com/piranna/send/tree/index) merged upstream so we can use upstream [serve-static](https://github.com/piranna/serve-static) on [Davius](https://github.com/piranna/Davius). I have open [two pull-requests](https://github.com/pillarjs/send/pulls/piranna) that need to be reviewed, hope to have done soon, but if someone wants to give it a try please tell me.
- [ ] replace custom [forever](https://github.com/piranna/forever/tree/starter) and [forever-monitor](https://github.com/piranna/forever-monitor) by [PalmTree](https://github.com/lite20/PalmTree) (I was sure this was already done... :-/)
- [ ] automate check and update of newer versions of external dependencies like `libfuse` or the Linux kernel using [buho](https://github.com/piranna/buho).
- [ ] update [pstree](https://github.com/piranna/pstree) to use upstream [scanf](https://github.com/piranna/node-scanf)
- [ ] update [download-manager](https://github.com/piranna/download-manager) to use upstream new [download](https://github.com/NodeOS/download) 4.x. This is specially important since they have added streamed downloads, making it to require a lot less memory by doing the decompression on the fly

**Comment by Ryuno-Ki** [comment]
Since NodeOS is managed per npm, I wonder, whether the process can be automated using [npm ls](https://docs.npmjs.com/cli/ls) … It has a `--json` and a `--parseable` flag.

**Comment by piranna** [comment]
> Since NodeOS is managed per npm, I wonder, whether the process can be automated using npm ls … It has a --json and a --parseable flag.

Yes it can, in fact `coverdeeps uses it, but now that we have done the dependencies tree is just a matter of maintain it updated.

**Comment by piranna** [comment]
Replaced `download` module from `download-manager` for `got`.

**Comment by aulvi** [comment]
@piranna I organize my local Node.js meetup and will see about making a short presentation about NodeOS in hopes of piquing interest in my local dev community. Next meetup is a couple of weeks away, I'll let you know how it goes.

**Comment by piranna** [comment]
That's cool! :-D Will it be recorded? We could put all the keynotes and
slides on the website for others to see :-)

**Comment by Coretool** [comment]
Hm... we've got a CCC (Chaos Computer Club) in Zurich, maybe one of them is interested, I'll check that ^^.

**Comment by piranna** [comment]
> Hm... we've got a CCC (Chaos Computer Club) in Zurich, maybe one of them
is interested, I'll check that ^^.

Wow, that would be an honor if you would do a talk about NodeOS on the
CCC... :-)

Here at Spain there used to be here some similar hacker conventions, but
they were mostly related with the okupy movement and now they have
dissapeared since there are no more okupied social centers here at Madrid
(or I'm not aware of them anymore...) :-/ Curious fact, Pablo Iglesias
(leader of Podemos political party) and me used to go 13 years ago to
"Laboratorio 3", the most important center of Madrid at that time :-P

**Comment by piranna** [comment]
I have [subscribed](https://becas.bifrutas.com/proyecto/117) NodeOS to be candidate for a bachelor to help develop the project and also get more visibility on TV and the media. According to rules the likes are not taken in account, but they help to improve visibility and at least we are on the first page :-) What I have doubts is about try to opt for one of the bachelor of up to 8000€ each month (so there are more posibilities) that would help to have someone working during some months or go to conferences like current one NodeInteractive, or go for the big one of 20000€ that would allow to have someone working for a year, but there's only one of them and will be given next year :-/ What do you think?

And on other things, I will have an introductory meeting with some investors next monday afternoon to talk about possible case of use and commercial posibilities for NodeOS, so any idea will be welcome :-)

**Comment by aulvi** [comment]
@piranna Commercial licensing is an interesting idea, I'm very curious to hear how your conversation with the investors go.

If you can provide to the investors a very clear picture of what the money will help you accomplish, all the better. At this stage in the project, the investors would be investing in YOU, so being confident and positive in your meeting should go a long way.

**Comment by piranna** [comment]
> @piranna Commercial licensing is an interesting idea

Could be, but probably this would be a re-license of the project, since it was discussed about use MIT license, so everybody is free to do comercial usage for the code. Using a more restrictive license like GPL would have open the door to a dual-license scheme for commercial usage, though. This could be a dificult thing to explain to them, but my intention is to left NodeOS and related project as Free Software, and if so, left as closed source only new ad-hoc modules, but definitely nothing in the core and if possible anything derived from it.


> I'm very curious to hear how your conversation with the investors go.

I'll post here the results. If it not get cancel or delayed, will be next monday at 15:00 Spain timezone, so hope to put the updates that night or next day in the morning.


> If you can provide to the investors a very clear picture of what the money will help you accomplish, all the better

At this moment the main problem would be my lack of time, while some months ago I had a lot of it... :-P But in this case I think this first conversation will be mostly about what commercial viability and use cases would NodeOS offer, not so much about money itself, just money-making ideas. So the question is, **what would you use NodeOS for?** :-)


> At this stage in the project, the investors would be investing in YOU, so being confident and positive in your meeting should go a long way

Thank you for the encouragment! :-D

**Comment by Ryuno-Ki** [comment]
Monday afternoon was yesterday. What was the outcome, @piranna?

**Comment by piranna** [comment]
> Monday afternoon was yesterday. What was the outcome, @piranna?

Lol :-P Not so bad. They are excited and impresed about the project and where taking a lof of notes about what I was telling them trying to figure out how it could be monetized, but the problem seems to be that NodeOS don't have a clearly defined target. Yes, it can be used on cloud servers, but the key question was, "why I would be interested on using NodeOS instead of <raw linux/Heroku/whatever>?". I explained them that people is trying to use it on mobile phones and the interest on use it as a desktop system (they think it could be a step backwards, by the way) and how much performance it has or how much difficult it could be to port it to ARM or other architectures. They could be somewhat interested on using it for IoT in "fog computing" (the space between the real hardware and online cloud) to provide low latency operation results for microcontroller-based IoT devices and sensors instead of doing the long trip to the cloud servers, so in this aspect they are mostly interested about CPU and battery usage benchmarks. We have agreed to keep in touch to see if in the new days we get some specific target where NodeOS could fit and there's no actual alternative in terms of performance or CPU and battery consumption. And of course, we should start trying to work hard to make NodeOS to run on ARM...

**Comment by aulvi** [comment]
@piranna NodeOS is a solution looking for a problem :)

Once you step into the IoT space, things makes a lot more sense. Buy some NodeOS-supported boards, use javascript to develop the entire solution, and quickly bring a product to market.

I believe the strength of the platform is developer accessibility, not necessarily performance or battery life. Marketing the project from that perspective could be a lot easier than attempting to compete with purpose-built low-power IoT devices.

**Comment by piranna** [comment]
> @piranna NodeOS is a solution looking for a problem :)

Like all my projects...


> Once you step into the IoT space, things makes a lot more sense. Buy some NodeOS-supported boards, use javascript to develop the entire solution, and quickly bring a product to market.

At this moment we only support x86, that leaves us with Intel Galileo and similar ones. We need to move to ARM and Raspberry Pi ASAP.


> I believe the strength of the platform is developer accessibility, not necessarily performance or battery life. Marketing the project from that perspective could be a lot easier than attempting to compete with purpose-built low-power IoT devices.

Obviously, but the feeling they gave me is that developer accessibility is not so much important as other economical factors, so that would left us in fast-paced environments or homebrew projects. Anyway I think it would be a good thing to focus on being developer friendly so later they could be interested on using it on production, and eating our own dog food would be a good thing too, maybe that last one point is the one that mostly motivates me at this moment to get some finantial backend, not so much about the money itself (we have lived until now without it with no problems) but instead to not being a solution looking for a problem anymore :-)


Disclaimer: yesterday my old company [Kurento was adquired by Twilio](http://www.kurento.org/blog/whats-next-kurento-and-elasticrtc) and I was not on the boat since a year ago (in part derived by personal disputes with my boss due to NodeOS), and I'm a bit depresive and pesimistic due to that, so don't take it too much in account on this or other messages.

**Comment by varinyc** [comment]
If you could port this to ARM, that would be astounding IMO. I always saw this project as an eventual means to run a server/js OS on IoT devices, was hoping to allow applications it's own little micro-web bubble in-between lots of tiny devices. So if say 3 devices running NodeOS all shared work to form a bigger picture, I'd be gamifying the devices into cool aspects of a web-based app game.

Simplest example: Pretend I'm making a sims game where YOU/I are the sim, I'm collecting gold on my phone from completing simple tasks such as: SmartWatch notices hunger ( by insert medical watch implementation), Alerts you via SIMS that fridge needs checked, gain a point for stocking fridge with key foods.

All devices are programmed in JS to handle tasks and manipulate data, but could use your phones power to generate the respective images and other GPU type things, and simply in-house pass them to the device that needs it to be there, but cannot handle it on it's own. OR doesn't HAVE to do it on it's own. That can all be packed within the game app, or game app server to app to other devices.

Respects to however this project turns out :) I'm just excited, and figured an outside look on what someone else thought it might have been, could be helpful.

**Comment by piranna** [comment]
Well, ARM port is WIP, the cross-compiler is able to generate ARM code and in fact the Linux kernel can be compiled, but it doesn't work. When I tried it almost one year ago I was able to make it boot on QEmu but got a kernel panic when giving control to the init process, but now I only get a black screen. I'm not sure if it's a problem of NodeOS or about QEmu, needs to do more checks on that area and I'm not an ARM expert (in fact, didn't used it before), so any help on that topic is welcome. Maybe it's a matter of include u-boot, don't know. The target at this moment is Raspberry Pi / raspi2, in case you are interested in give it a try.

**Comment by junland** [comment]
@piranna I'd like to add that maybe NodeOS might be suited for Hypervisors such as KVM, Xen, and VMware. NodeOS, kinda fits in the paradigm of "unikernel" seems like a there's a couple of projects like "MirageOS" (OCamel) and "IncludeOS" (C/C++) that have / working on achieving this with there respective programming languages. With that, NodeOS would be stepping into the Enterprise space, which might be a good outlet for funding / support from other companies / contributors.

**Comment by varinyc** [comment]
ARM uses Big.little technology, there are sets of little cores which you can utilize for general data manipulation and procedures, files etc., then generally a smaller set of Big cores that help function as busrty powerhouses used in visuals and other power consuming operations. If you keep a mindset that you can add balance in the change of things instead of each thing, that way utilizing all cores is easier and more async.

**Comment by piranna** [comment]
> @piranna I'd like to add that maybe NodeOS might be suited for Hypervisors such as KVM, Xen, and VMware. NodeOS, kinda fits in the paradigm of "unikernel" seems like a there's a couple of projects like "MirageOS" (OCamel) and "IncludeOS" (C/C++) that have / working on achieving this with there respective programming languages. With that, NodeOS would be stepping into the Enterprise space, which might be a good outlet for funding / support from other companies / contributors.

Not a bad idea, this is mostly a matter of how to configure the kernel, do you know how to do and test this? In this point, now that we have Docker support probably would be of more interest for the enterprise environment, isn't it?

**Comment by piranna** [comment]
For the purposse of this issue, I have found [this list](https://medium.freecodecamp.com/how-we-got-a-2-year-old-repo-trending-on-github-in-just-48-hours-12151039d78b#.12o4e5igs) of points about how to promote an open source project. Probably it would be a good idea to think what of them we would implement and how, what do you think?
